### DISCUSSION TITLE
ROS-Compatibility

### DISCUSSION THREAD

**Discussion by abuibaid** [google-deepmind/mujoco#990] [General]
Hello everyone! As a long-time user of Mujoco, I believe that creating a bridge between Mujoco and ROS would be an amazing advancement. The combination of Mujoco's accuracy and the communication capabilities of ROS would result in a powerful tool that developers would have for sure use it. While there are existing repositories offering this integration, having an official Mujoco-supported bridge would be highly desirable. Regards, Khalil

**Comment by yuvaltassa** [comment]
We agree and are looking into options to make this happen. Can you recommend someone who could be contracted to do this?

**Reply by DavidPL1** [reply]
We at Bielefeld University have been developing a MuJoCo based simulation with ROS  (https://github.com/ubi-agni/mujoco_ros_pkgs) over the past year.

The idea was to seamlessly extend MuJoCo and provide a similar, extendable interface as Gazebo does. The current state supports a pluginlib based plugin-system for which we already provide a [plugin](https://github.com/ubi-agni/mujoco_ros_pkgs/tree/noetic-devel/mujoco_ros_sensors) to automatically stream out most of the current sensors, a [plugin](https://github.com/ubi-agni/mujoco_ros_pkgs/tree/noetic-devel/mujoco_ros_sensors) that enables ROS Control support to steer any Robot where matching URDF and MuJoCo models are provided and we are working on simulation of tactile sensors based on soft contacts.

some more features are:
- run headless or with UI (heavily based on MuJoCo Simulate)
- ROS services to get and set the current simulation state (pause, unpause, reload, reset, object position/velocitiy/properties, ...)
- RGB-D + segmentation image camera streams with automatic camera_info topics

![mujoco_ros](https://github.com/deepmind/mujoco/assets/20286447/29bddc8e-2a3f-4012-8197-d3dc0654b0b1)


I am also currently working on [extending the official python bindings](https://github.com/deepmind/mujoco/issues/983?notification_referrer_id=NT_kwDOATWL77M3MDUxNjY0Mjk1OjIwMjg2NDQ3#issuecomment-1647576306https://github.com/ubi-agni/mujoco_ros_pkgs/tree/noetic-devel/mujoco_ros_sensors) to allow combined usage of ROS, MuJoCo and any ML/DL framework in python.

The whole project is under active development, and there isn't as much documentation yet.
[Here](https://github.com/ubi-agni/mujoco_ros_demos) is a repository with some examples (Gazebo's rrbot ported to mujoco_ros and the Franka Emika Panda controlled via ROS Control)


We'd be happy to work closer with MuJoCo's developers and community on this and welcome contributions.

**Reply by abuibaid** [reply]
Hi again. Yes, I could recommend @DavidPL1 and @HoangGiang93. They have done really good approaches. Easy to install, their repositories are maintainable and updatable.

**Reply by HoangGiang93** [reply]
Thank you, @abuibaid, for your recommendation. I've also developed the ROS interface for MuJoCo, which you can find in the following repository: https://github.com/HoangGiang93/mujoco_sim. My main focus in this ROS interface is to leverage the MuJoCo physics engine for optimal physics interactions. Here are some key features of my approach:

- Integrates controller interfaces, controller managers, and hardware interfaces sourced from http://wiki.ros.org/ros_control
- Implements PD computed-torque control for enhanced stability
- Enables spawning and destruction of objects during run-time via rosservice
- Synchronizes simulation time with real time, with the ability to adjust simulation time for faster or slower execution
- Provides comprehensive visualization of MuJoCo elements within rviz
- Offers velocity control for the base

**Comment by sunzhon** [comment]
I developed a lightweight ROS interface called [mujoco_ros](https://github.com/sunzhon/mujoco_ros.git), which can be integrated with any version of MuJoCo without modifying its code or CMake configurations. This interface allows MuJoCo to function as a ROS node (e.g., /mujoco_sim), transferring the simulated robot state to a ROS topic (e.g., /ambot_v2/states)  and executing commands from another ROS topic (e.g.,/ambot_v2/action ) generated by a custom controller ROS Node (e.g., /controller_node).
![image](https://github.com/user-attachments/assets/67d75bdc-c8f7-4b6c-a0c4-0eb528b45fea)

**Comment by GabrielCosme** [comment]
If you only need to get sensor data and set the actuators, there is a [plug and play plugin for using ROS2 with MuJoCo](https://github.com/Team-Micras/mujoco_ros2_plugin), it is only needed to install the plugin and add its instance to the XML model file:

```xml
<extension>
  <plugin plugin="mujoco.ros2">
    <instance name="ros2_plugin">
  </plugin>
</extension>
```

And it will automatically publish data for all the sensors in separate topics, and create subscribers for the actuators.

**Comment by Noe-Sanchez** [comment]
I dont wish to necro, but has there been any official update or releases for this @yuvaltassa ?

**Reply by abuibaid** [reply]
The ros_controlls community are maintaing one for ros2. It is called mujoco_ros2_control
